x/evm/keeper/vote_handler.go

```diff
@@ -12,65 +12,153 @@ import (
-		}
+func (v voteHandler) IsFalsyResult(result codec.ProtoMarshaler) bool {
+	voteEvents, err := types.UnpackEvents(v.cdc, result.(*vote.Vote).Result)
 
-		chain, ok := nexus.GetChain(ctx, voteEvents.Chain)
-		if !ok {
-			return fmt.Errorf("%s is not a registered chain", voteEvents.Chain)
-		}
+	return err != nil || len(voteEvents.Events) == 0
+}
+
+func (v voteHandler) HandleExpiredPoll(ctx sdk.Context, poll vote.Poll) error {
+	rewardPoolName, ok := poll.GetRewardPoolName()
+	if !ok {
+		return fmt.Errorf("reward pool not set for poll %s", poll.GetKey().String())
+	}
 
-		if !keeper.HasChain(ctx, voteEvents.Chain) {
-			return fmt.Errorf("%s is not an evm chain", voteEvents.Chain)
+	// TODO: MarkChainMaintainerMissingVote for those who didn't vote in time. Need
+	// to be able to get chain of expired polls in order to do it.
+	rewardPool := v.rewarder.GetPool(ctx, rewardPoolName)
+	// Penalize voters who failed to vote
+	for _, voter := range poll.GetVoters() {
+		if !poll.HasVoted(voter.Validator) {
+			rewardPool.ClearRewards(voter.Validator)
+			v.keeper.Logger(ctx).Debug("penalized voter due to timeout",
+				"voter", voter.Validator.String(),
+				"poll", poll.GetKey().String())
 		}
+	}
 
-		for _, voter := range poll.GetVoters() {
-			hasVoted := poll.HasVoted(voter.Validator)
-			hasVotedIncorrectly := hasVoted && !poll.HasVotedCorrectly(voter.Validator)
+	return nil
+}
 
-			nexus.MarkChainMaintainerMissingVote(ctx, chain, voter.Validator, !hasVoted)
-			nexus.MarkChainMaintainerIncorrectVote(ctx, chain, voter.Validator, hasVotedIncorrectly)
-		}
+func (v voteHandler) HandleCompletedPoll(ctx sdk.Context, poll vote.Poll) error {
+	voteEvents, err := types.UnpackEvents(v.cdc, poll.GetResult().(*vote.Vote).Result)
+	if err != nil {
+		return err
+	}
 
-		if len(voteEvents.Events) == 0 {
-			poll.AllowOverride()
+	chain, ok := v.nexus.GetChain(ctx, voteEvents.Chain)
+	if !ok {
+		return fmt.Errorf("%s is not a registered chain", voteEvents.Chain)
+	}
 
-			return nil
-		}
+	rewardPoolName, ok := poll.GetRewardPoolName()
+	if !ok {
+		return fmt.Errorf("reward pool not set for poll %s", poll.GetKey().String())
+	}
+
+	rewardPool := v.rewarder.GetPool(ctx, rewardPoolName)
+
+	for _, voter := range poll.GetVoters() {
+		hasVoted := poll.HasVoted(voter.Validator)
+		hasVotedLate := poll.HasVotedLate(voter.Validator)
+		hasVotedIncorrectly := hasVoted && !poll.HasVotedCorrectly(voter.Validator)
 
-		chainK := keeper.ForChain(chain.Name)
-		cacheCtx, writeCache := ctx.CacheContext()
+		v.nexus.MarkChainMaintainerMissingVote(ctx, chain, voter.Validator, !hasVoted)
+		v.nexus.MarkChainMaintainerIncorrectVote(ctx, chain, voter.Validator, hasVotedIncorrectly)
 
-		err = handleEvents(cacheCtx, chainK, voteEvents.Events, chain)
-		if err != nil {
-			// set events to failed, we will deal with later
-			for _, e := range voteEvents.Events {
-				chainK.SetFailedEvent(ctx, e)
+		v.keeper.Logger(ctx).Debug("marked voter misbehave",
```

Comment:
```suggestion
		v.keeper.Logger(ctx).Debug("marked voter behaviour",
```
Also, similarly, would be nice to add voter in the msg content.